Handle IPv4 validation with leading zeros on go 1.17

orelmisan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
Since go version 1.17, net.ParseIP() and net.ParseCIDR() functions rejects leading zeros in the dot-decimal notation of IPv4 addresses.
In order to upgrade Kubevirt's go version to 1.17 and to maintain backwards compatibility - we need to use ParseIPSloppy() and ParseCIDRSloppy() functions from k8s.io/utils.
The affected fields in the VMI spec are:
- dnsConfig.nameservers (k8s.io.api.core.v1.PodDNSConfig)
- domain.devices.interfaces.dhcpOptions.ntpServers (v1.DHCPOptions)

These fields will always be parsed as decimal IPv4 addresses (leading zeros are ignored):
For example: 192.168.010.1 => 192.168.10.1

This PR:
1. Lets the above fields to be validated successfully on go 1.17.
2. Adds E2E test cases to make sure the above fields are backward compatible.

For additional details please see issue #6498.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
This change was tested locally on go 1.17 - the above fields are passing validation.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

orelmisan commented:
Change: Remove not applicable unit test.

orelmisan commented:
/cc @EdDev @ormergi

EdDev commented on tests/network/vmi_networking.go:
```diff
@@ -511,6 +511,34 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
 			Expect(err).ToNot(HaveOccurred())
 		})
+
+		It("[test_id:xxxx]should allow NTP servers' IPv4 addresses which contain leading zeros", func() {
```
First, you do not need the `test_id`, please drop it.

Secondly, could you please clarify what is the advantage to have another dedicated e2e test for this edge case? e2e tests in general are costly because the creation and termination of a VMI takes a significant time.

How about just add to the prev test a 3rd address, e.g.: `172.0.0.099` ? You can place it in a const and explain in a comment above it that it is intentional.

orelmisan replied:
I don't think we have to keep a dedicated test for this case.
Although it would be nice to have a focused warning when this specific feature fails, I think the cost is too high for the value it could provide.

It was a response to @ormergi 's comment: https://github.com/kubevirt/kubevirt/pull/6986#pullrequestreview-840082503

EdDev commented on tests/network/vmi_networking.go:
```diff
@@ -572,6 +572,32 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
 			Expect(err).ToNot(HaveOccurred())
 		})
+
+		It("[test_id:1779]should allow parsing of DNS nameservers' IPv4 addresses with leading zeros from resolv.conf", func() {
```
Same here.

orelmisan replied:
See comment above.

EdDev reviewed (commented):
Commit messages:

>  Upgrade k8s.io/utils module 
go.mod: Upgrade k8s.io/utils module

For all the other commits, please start the topic with lowercase letter (e.g. `network: `).

orelmisan commented:
> Commit messages:
> 
> > Upgrade k8s.io/utils module
> > go.mod: Upgrade k8s.io/utils module
> 
> For all the other commits, please start the topic with lowercase letter (e.g. `network: `).
@EdDev Done.

orelmisan commented:
Changes:
1. Fixed commit subjects following review.
2. Removed dedicated E2E tests.
3. Added test cases to existing E2E tests for NTP servers and DNS nameservers.

EdDev commented on tests/network/vmi_networking.go:
```diff
@@ -525,8 +525,13 @@ var _ = SIGDescribe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:c
+			// This IPv4 address tests backwards compatibility of the "DNSConfig.Nameservers" field.
+			// The leading zero is intentional.
+			// For more details please see: https://github.com/kubevirt/kubevirt/issues/6498
+			const DNSNameserverWithLeadingZeros = "01.1.1.1"
```
I think this will pass even on 1.17, as it `01` is `1` and in the right range. Also, consider using a loopback address so no surprises will occur. I do not know why the current addresses used are public ones.

orelmisan replied:
Without this PR, on go 1.17 - this address will fail on the validating webhook with the following error:
```
$ ./cluster-up/kubectl.sh apply -f examples/vmi-fedora.yaml 
...
The request is invalid: 
* nameservers: must be valid IP address: 01.1.1.1
```

With this PR, on go 1.17 - this address will be accepted, but will fail when reading /etc/resolv.conf from virt-launcher's pod (for using it as an input for the DHCP server).

As for your second comment, the addresses here don't really matter because we only test that the DHCP server had sent the VM a proper address, and then the VM is terminated.

orelmisan replied:
BTW, 1.1.1.1 is Cloudflare's DNS server

EdDev reviewed (commented):
Looks great, one small issue inline.

orelmisan commented:
Change: 
1. Added a utility function IsIPv4Address
2. Used it on VMI validating webhook

orelmisan commented:
Change: Fixed a typo in a commit message

EdDev commented on pkg/util/net/ip/ip.go:
```diff
@@ -83,3 +83,17 @@ func isIPv6Disabled(filename string) bool {
 	return data[0] == '1'
 }
+
+func IsIPv4Address(address string, useSloppyParsing bool) bool {
```
Ahh.. we are back a few steps now.

- Using a boolean as an argument is always a smell, and here it is obvious. This function does two things and the caller controls it. So there is no reason to have one function.. there should be two. But if there are two, you do not really need it, right?
- I prefer no one will use util/net/ip package now that we have the network package available. We need to move everything out of here.. or at least the ones we care about.

EdDev reviewed (changes requested):
All is good except the `IsIPv4Address` commit which is not needed IMO.
I think it makes it less readable and adds uneeded code.

orelmisan commented:
> Change:
> 
> 1. Added a utility function IsIPv4Address
> 2. Used it on VMI validating webhook
> 
> EDIT: Answers @ormergi's comment [#6986 (review)](https://github.com/kubevirt/kubevirt/pull/6986#pullrequestreview-840082503)

@EdDev reverted this change.

EdDev reviewed (approved):
Thank you!

orelmisan commented:
/retest-required

EdDev commented:
/sig network

orelmisan commented:
rebase (to pass pull-kubevirt-generate test)
@EdDev can you please lgtm again?

orelmisan commented:
/retest-required

orelmisan commented:
Work continues on PR #6986.
